fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/api/transfer/v1/storage/Storage.java

```diff
@@ -0,0 +1,203 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.transfer.v1.storage;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import org.jetbrains.annotations.Nullable;
+
+import net.fabricmc.fabric.api.transfer.v1.transaction.Transaction;
+import net.fabricmc.fabric.impl.transfer.TransferApiImpl;
+
+/**
+ * An object that can store resources.
+ *
+ * <p><ul>
+ *     <li>{@link #supportsInsertion} and {@link #supportsExtraction} can be used to tell if insertion and extraction
+ *     functionality are possibly supported by this storage.</li>
+ *     <li>{@link #insert} and {@link #extract} can be used to insert or extract resources from this storage.</li>
+ *     <li>{@link #iterator}, {@link #anyView} and {@link #exactView} can be used to inspect the contents of this storage.
+ *     Only one of these functions may be called for a given transaction parameter,
+ *     to ensure storages have flexibility in how they operate.</li>
+ *     <li>{@link #getVersion()} can be used to quickly check if a storage has changed, without having to rescan its contents.</li>
+ * </ul>
+ *
+ * <p><b>Important note:</b> Unless otherwise specified, all transfer functions take a non-empty resource
+ * and a non-negative maximum amount as parameters.
+ * Implementations are encouraged to throw an exception if these preconditions are violated.
+ *
+ * <p>For transfer functions, the returned amount must be non-negative, and smaller than the passed maximum amount.
+ * Consumers of these functions are encourage to throw an exception if these postconditions are violated.
+ *
+ * @param <T> The type of the stored resources.
+ * @see Transaction
+ */
+public interface Storage<T> {
+	/**
+	 * Return false if calling {@link #insert} will absolutely always return 0, or true otherwise or in doubt.
+	 *
+	 * <p>Note: This function is meant to be used by pipes or other devices that can transfer resources to know if
+	 * they should render a visual connection to this storage.
+	 */
+	default boolean supportsInsertion() {
+		return true;
+	}
+
+	/**
+	 * Try to insert up to some amount of a resource into this storage.
+	 *
+	 * @param resource The resource to insert. May not be empty.
+	 * @param maxAmount The maximum amount of resource to insert. May not be negative.
+	 * @param transaction The transaction this operation is part of.
+	 * @return A nonnegative integer not greater than maxAmount: the amount that was inserted.
+	 */
+	long insert(T resource, long maxAmount, Transaction transaction);
```

Comment:
`long insert(Transaction transaction, T resource, long amount)` > this

Reply:
The position of the transaction doesn't really matter. I prefer `maxAmount` because well it's the max amount, the real amount is what is returned.

Reply:
Marking as resolved, let me know if it's a major issue. (I doubt it is).